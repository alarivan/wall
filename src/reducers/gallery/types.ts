import { TGrid } from '../../types';
import { Map } from 'immutable';

export type TGalleryState = Map<string, TGrid>;

export const SAVE = 'SAVE';
export type SAVE_ACTION = {
  type: typeof SAVE;
  payload: {
    grid: TGrid;
  };
};

export const DELETE = 'DELETE';
export type DELETE_ACTION = {
  type: typeof DELETE;
  payload: {
    id: string;
  };
};

export type TGalleryAction = SAVE_ACTION | DELETE_ACTION | { type: 'EMPTY' };
