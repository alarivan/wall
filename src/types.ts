export type TColor = string | null;

export type TGrid = {
  meta: {
    rows: number;
    columns: number;
    width: number;
  };
  data: TColor[];
};

export const PAINT = 'PAINT';
export type PAINT_ACTION = {
  type: typeof PAINT;
  payload: {
    index: number;
    color: TColor;
  };
};

export const CLEAR = 'CLEAR';
export type CLEAR_ACTION = {
  type: typeof CLEAR;
};

export type TAction = PAINT_ACTION | CLEAR_ACTION | { type: 'EMPTY' };
