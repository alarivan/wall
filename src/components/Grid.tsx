import React from 'react';
import Cell from './Cell';
import StyledGrid from './StyledGrid';
import { TGrid, TColor } from '../types';

interface GridProps {
  grid: TGrid;
  brush: TColor;
  preview: boolean;
  onCellClick: (index: number, brush: TColor, color: TColor) => void;
}

const Grid: React.FC<GridProps> = ({ grid, brush, preview, onCellClick }) => {
  return (
    <StyledGrid {...grid.meta} preview={preview}>
      {grid.data.map((color, index) => (
        <Cell
          data-testid='grid-cell'
          onClick={() => onCellClick(index, brush, color)}
          key={index}
          border={!preview}
          cellColor={color}
        />
      ))}
    </StyledGrid>
  );
};

export default Grid;
