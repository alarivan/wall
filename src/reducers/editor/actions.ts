import { TColor } from '../../types';
import {
  PAINT_ACTION,
  PAINT,
  UPDATE_BACKGROUND_ACTION,
  UPDATE_BACKGROUND,
  CLEAR_ACTION,
  CLEAR,
  UPDATE_ROWS_ACTION,
  UPDATE_ROWS,
  UPDATE_COLUMNS_ACTION,
  UPDATE_COLUMNS,
  UPDATE_SIZE_ACTION,
  UPDATE_SIZE,
  UNDO_ACTION,
  UNDO,
  REDO_ACTION,
  REDO,
  RESET_ACTION,
  RESET,
  TState,
  SET_STATE_ACTION,
  SET_STATE,
} from './types';

export function paintAction(index: number, color: TColor): PAINT_ACTION {
  return {
    type: PAINT,
    payload: { index, color },
  };
}

export function updateBackgroundAction(
  color: TColor,
): UPDATE_BACKGROUND_ACTION {
  return {
    type: UPDATE_BACKGROUND,
    payload: { color },
  };
}

export function clearAction(): CLEAR_ACTION {
  return {
    type: CLEAR,
  };
}

export function updateRowsAction(value: number): UPDATE_ROWS_ACTION {
  return {
    type: UPDATE_ROWS,
    payload: { value },
  };
}

export function updateColumnsAction(value: number): UPDATE_COLUMNS_ACTION {
  return {
    type: UPDATE_COLUMNS,
    payload: { value },
  };
}

export function updateSizeAction(value: number): UPDATE_SIZE_ACTION {
  return {
    type: UPDATE_SIZE,
    payload: { value },
  };
}

export function undoAction(): UNDO_ACTION {
  return {
    type: UNDO,
  };
}

export function redoAction(): REDO_ACTION {
  return {
    type: REDO,
  };
}

export function resetAction(): RESET_ACTION {
  return {
    type: RESET,
  };
}

export function setStateAction(value: TState): SET_STATE_ACTION {
  return {
    type: SET_STATE,
    payload: {
      value,
    },
  };
}
