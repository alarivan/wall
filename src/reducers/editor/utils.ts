import uuidv4 from 'uuidv4';
import {
  TGridMeta,
  TGridMetaInit,
  TGridData,
  TGrid,
  TColor,
} from '../../types';
import { TState } from './types';

export function initMeta(meta: TGridMeta | TGridMetaInit): TGridMeta {
  return {
    id: meta.id || uuidv4(),
    columns: meta.columns,
    rows: meta.rows,
    size: meta.size,
    length: meta.columns * meta.rows,
    background: meta.background,
  };
}

export function initGrid(
  meta: TGridMeta | TGridMetaInit,
  data: TGridData = {},
): TGrid {
  return {
    meta: initMeta(meta),
    data,
    [Symbol.iterator]() {
      const length = this.meta.length;
      const data = this.data;
      let index = 0;

      return {
        next(): IteratorResult<TColor> {
          if (index >= length) {
            return {
              value: undefined,
              done: true,
            };
          }

          return {
            value: data[index++],
            done: false,
          };
        },
      };
    },
  };
}

export function initState(grid: TGrid): TState {
  return {
    history: [grid],
    current: 0,
  };
}
