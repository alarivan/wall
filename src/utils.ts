import { TGrid, TColor } from './types';

export function gridMap(
  grid: TGrid,
  callback: (item: TColor, index: number) => any,
): Element[] {
  let result = [];
  let index = 0;
  for (const item of grid) {
    result.push(callback(item, index++));
  }
  return result;
}
