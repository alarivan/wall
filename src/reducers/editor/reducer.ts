import {
  TState,
  TAction,
  PAINT,
  UPDATE_BACKGROUND,
  CLEAR,
  UPDATE_ROWS,
  UPDATE_COLUMNS,
  UPDATE_SIZE,
  UNDO,
  REDO,
  RESET,
  SET_STATE,
  PAINT_ACTION,
  UPDATE_BACKGROUND_ACTION,
  UPDATE_SIZE_ACTION,
  UPDATE_ROWS_ACTION,
  UPDATE_COLUMNS_ACTION,
} from './types';
import { initState, initGrid, initMeta } from './utils';
import { DEFAULT_BACKGROUND } from '../../constants';
import { TGrid, TGridData } from '../../types';

export const initialState: TState = initState(
  initGrid(
    { rows: 20, columns: 20, size: 20, background: DEFAULT_BACKGROUND },
    {},
  ),
);

export const gridReducer = (state: TState, action: TAction): TState => {
  switch (action.type) {
    case PAINT:
      return _updateGrid(state, _paintCell(action));

    case UPDATE_BACKGROUND:
      return _updateGrid(state, _updateBackground(action));

    case CLEAR:
      return _updateGrid(state, _clearGrid);

    case UPDATE_ROWS: {
      return _updateGrid(state, _updateRows(action));
    }

    case UPDATE_COLUMNS: {
      return _updateGrid(state, _updateColumns(action));
    }

    case UPDATE_SIZE:
      return _updateGrid(state, _updateSize(action));

    case UNDO:
      if (state.current > 0) {
        return Object.assign({}, state, { current: state.current - 1 });
      }
      return state;

    case REDO:
      if (state.current < state.history.length - 1) {
        return Object.assign({}, state, { current: state.current + 1 });
      }
      return state;

    case RESET:
      return initialState;

    case SET_STATE:
      return action.payload.value;

    default:
      return state;
  }
};

function _updateGrid(state: TState, fn: (grid: TGrid) => TGrid): TState {
  const currentHistory = state.history.slice(0, state.current + 1);
  const current = currentHistory[state.current];
  const grid = fn(current);

  return _addToHistory(currentHistory, grid, state);
}

function _addToHistory(history: TGrid[], next: TGrid, state: TState): TState {
  return Object.assign({}, state, {
    history: [...history, next],
    current: state.current + 1,
  });
}

function _clearGrid(grid: TGrid): TGrid {
  return initGrid(grid.meta);
}

function _paintCell(action: PAINT_ACTION): (grid: TGrid) => TGrid {
  return function(grid: TGrid): TGrid {
    const index = action.payload.index;
    return Object.assign({}, grid, {
      data: { ...grid.data, [index]: action.payload.color },
    });
  };
}

function _updateBackground(
  action: UPDATE_BACKGROUND_ACTION,
): (grid: TGrid) => TGrid {
  return function(grid: TGrid): TGrid {
    return Object.assign({}, grid, {
      meta: {
        ...grid.meta,
        background: action.payload.color,
      },
    });
  };
}

function _updateSize(action: UPDATE_SIZE_ACTION): (grid: TGrid) => TGrid {
  return function(grid: TGrid): TGrid {
    return Object.assign({}, grid, {
      meta: { ...grid.meta, size: action.payload.value },
    });
  };
}

function _updateRows(action: UPDATE_ROWS_ACTION): (grid: TGrid) => TGrid {
  return function(grid: TGrid): TGrid {
    const difference: number = action.payload.value - grid.meta.rows;

    let data: TGridData = {};
    for (let [key, value] of Object.entries(grid.data)) {
      const intKey: number = parseInt(key);
      const lastLineFirstIndex =
        grid.meta.length - grid.meta.columns * Math.abs(difference);

      if (difference > 0 || intKey < lastLineFirstIndex) {
        data[key] = value;
      }
    }

    return Object.assign({}, grid, {
      data,
      meta: initMeta({
        ...grid.meta,
        rows: action.payload.value,
      }),
    });
  };
}

function _updateColumns(action: UPDATE_COLUMNS_ACTION): (grid: TGrid) => TGrid {
  return function(grid: TGrid): TGrid {
    const difference: number = action.payload.value - grid.meta.columns;

    let data: TGridData = {};
    for (let [key, value] of Object.entries(grid.data)) {
      const intKey: number = parseInt(key);
      const line = _getLine(grid.meta.columns, intKey);
      const edge = _getEdge(difference, grid.meta.columns, line);
      if (difference > 0 || !(intKey >= edge)) {
        const index: string = (intKey + difference * line).toString();
        data[index] = value;
      }
    }

    return Object.assign({}, grid, {
      data,
      meta: initMeta({
        ...grid.meta,
        columns: action.payload.value,
      }),
    });
  };
}

function _getLine(columns: number, index: number): number {
  const line = Math.floor(index / columns) - 1;
  return line + 1;
}

function _getEdge(difference: number, columns: number, line: number): number {
  return line * columns + (columns - Math.abs(difference));
}
