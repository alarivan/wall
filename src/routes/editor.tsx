import React, { useContext, useReducer, useEffect } from 'react';
import Container from '../components/Container';
import { RouteComponentProps } from '@reach/router';
import { GalleryContext } from '../GalleryContext';
import { TGalleryState } from '../reducers/gallery/types';
import { initState } from '../reducers/editor/utils';
import { initialState, gridReducer } from '../reducers/editor/reducer';
import { setStateAction } from '../reducers/editor/actions';

interface Props extends RouteComponentProps {
  id?: string;
}

function _getGrid(id: string | undefined, state: TGalleryState) {
  const grid = id ? state.get(id) : undefined;

  return grid ? initState(grid) : initialState;
}

const Editor: React.FC<Props> = ({ id }) => {
  const { state: galleryState } = useContext(GalleryContext);
  const [state, dispatch] = useReducer(gridReducer, _getGrid(id, galleryState));

  useEffect(() => {
    dispatch(setStateAction(_getGrid(id, galleryState)));
  }, [id, galleryState]);

  return <Container state={state} dispatch={dispatch} />;
};

export default Editor;
