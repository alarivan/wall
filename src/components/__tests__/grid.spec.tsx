import React from 'react';
import { render, cleanup, fireEvent } from '@testing-library/react';
import { TGrid } from '../../types';
import Grid from '../Grid';
import { initGrid } from '../../reducers/editor/utils';

afterEach(cleanup);

const initialState: TGrid = initGrid({
  id: 'fake_id',
  columns: 2,
  rows: 2,
  size: 10,
  length: 4,
  background: 'red',
});

describe('Grid', () => {
  it('renders correctly', () => {
    const { asFragment } = render(
      <Grid
        grid={initialState}
        onCellClick={jest.fn()}
        preview={false}
        brush='red'
      />,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('call onCellClick', () => {
    const props = {
      grid: initialState,
      onCellClick: jest.fn(),
      preview: false,
      brush: 'red',
    };
    const { queryAllByTestId } = render(<Grid {...props} />);

    fireEvent.click(queryAllByTestId('grid-cell')[0]);

    expect(props.onCellClick).toHaveBeenCalledTimes(1);
  });
});
