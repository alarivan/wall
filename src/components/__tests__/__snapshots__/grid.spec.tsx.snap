// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grid renders correctly 1`] = `
<DocumentFragment>
  .c1 {
  background: transparent;
  border-style: solid;
  border-color: black;
  border-width: 0 1px 1px 0;
  -webkit-transition: -webkit-transform 100ms ease-in-out;
  -webkit-transition: transform 100ms ease-in-out;
  transition: transform 100ms ease-in-out;
  cursor: pointer;
}

.c1:hover {
  border-width: 1px;
  -webkit-transform: scale(1.1);
  -ms-transform: scale(1.1);
  transform: scale(1.1);
}

.c0 {
  display: grid;
  -webkit-box-pack: start;
  -webkit-justify-content: start;
  -ms-flex-pack: start;
  justify-content: start;
  grid-gap: 0;
  grid-template-columns: repeat( 2,10px );
  grid-template-rows: repeat(2,10px);
  border-style: solid;
  border-width: 1px;
  border-color: black white white black;
  background: #c2c2d1;
}

<div
    class="c0"
    rows="2"
    size="10"
  >
    <div
      class="c1"
      data-testid="grid-cell"
    />
    <div
      class="c1"
      data-testid="grid-cell"
    />
    <div
      class="c1"
      data-testid="grid-cell"
    />
    <div
      class="c1"
      data-testid="grid-cell"
    />
  </div>
</DocumentFragment>
`;
